src/odemis/driver/test/focustracker_test.py

```diff
@@ -0,0 +1,187 @@
+        self.fake_focus_tracker.tracking.value = False
+        self.assertFalse(self.fake_focus_tracker.tracking.value)
+        self.assertFalse(self.fake_focus_tracker.get_tracking())
+
+    def test_pid_gains(self):
+        """Test that PID gains are set when within range and remain unchanged when new gain is out of range."""
+        new_p = 0.5
+        new_i = 15
+        new_d = 3
+        self.fake_focus_tracker.updateMetadata({model.MD_GAIN_P: new_p, model.MD_GAIN_I: new_i, model.MD_GAIN_D: new_d})
+        set_p = self.fake_focus_tracker.get_proportional()
+        self.assertEqual(new_p, set_p)
+        set_i = self.fake_focus_tracker.get_integral()
+        self.assertEqual(new_i, set_i)
+        set_d = self.fake_focus_tracker.get_derivative()
+        self.assertEqual(new_d, set_d)
+
+
+class TestFocusTrackerCO(unittest.TestCase):
+    """Test the focus tracker functionality with the CANopen network."""
+    backend_was_running = False
+
+    @classmethod
+    def setUpClass(cls):
+        try:
+            test.start_backend(FOCUS_TRACKER_CONFIG)
+        except LookupError:
+            logging.info("A running backend is already found, skipping tests")
+            cls.backend_was_running = True
+            return
+        except IOError as exp:
+            logging.error(str(exp))
+            raise
+
+        # find components by their role
+        cls.focus_tracker = model.getComponent(role="focus-tracker")
+
+    @classmethod
+    def tearDownClass(cls):
+        if cls.backend_was_running:
+            return
+        test.stop_backend()
+
+    def test_connection(self):
+        """Test that connection to the network is successful when connected, and unsuccessful when disconnected."""
+        # If the network scanner does not see a node at index 0x10, the focus tracker is not connected.
+        self.assertTrue(0x10 in self.focus_tracker.get_available_nodes())
+        # If a fake channel name is entered, an IOError should be raised.
+        with self.assertRaises(IOError):
+            FocusTrackerCO(name="Focus Tracker", role="focus_tracker", channel='not a channel', node_idx=0x10)
+        # When the node_idx is incorrect an empty node is added to the network, this then results in an
+        # SDOCommunicationError when we try to send a message to an object in the object dictionary.
+        with self.assertRaises(SdoCommunicationError):
+            FocusTrackerCO(name="Focus Tracker", role="focus_tracker", channel='can0', node_idx=0x12)
+
+    def test_object_dictionary(self):
+        """Verify that all relevant objects are in the object dictionary."""
+        object_names = ["AI Input PV", "CO Set Point W", "CO Proportional Band Xp1",
+                        "CO Integral Action Time Tn1", "CO Derivative Action Time Tv1"]
+        object_dictionary = self.focus_tracker.get_object_dictionary()
+        self.assertTrue(all(n in object_dictionary.names for n in object_names))
```

Comment:
I don't really think this test is necessary. That's all the internal business of the driver. As long as it behaves properly, it the actual names of the canOpen dict shoudn't matter.